File: pi-client/media_player.py
#!/usr/bin/env python3
"""
Mesophy Pi Client - Native Media Player
Hardware-accelerated media playback using Pi native tools
"""

import os
import json
import time
import subprocess
import threading
import queue
from pathlib import Path
from PIL import Image

class NativeMediaPlayer:
    def __init__(self, content_dir='/opt/mesophy/content', config_path='/opt/mesophy/config/config.json'):
        self.content_dir = Path(content_dir)
        self.config_path = config_path
        self.config = self.load_config()
        
        # Display settings
        self.display_width = self.config.get('display', {}).get('width', 1920)
        self.display_height = self.config.get('display', {}).get('height', 1080)
        
        # Media player processes
        self.current_process = None
        self.playlist_queue = queue.Queue()
        self.playback_thread = None
        self.is_playing = False
        
        # Supported formats
        self.video_formats = {'.mp4', '.mov', '.avi', '.mkv', '.webm'}
        self.image_formats = {'.jpg', '.jpeg', '.png', '.gif', '.bmp', '.webp'}
        
        # Create content directory
        self.content_dir.mkdir(parents=True, exist_ok=True)
        
        print("Native Media Player initialized")

    def load_config(self):
        """Load configuration"""
        try:
            if os.path.exists(self.config_path):
                with open(self.config_path, 'r') as f:
                    return json.load(f)
            return {}
        except Exception as e:
            print(f"Error loading config: {e}")
            return {}

    def stop_current_playback(self):
        """Stop any currently playing media"""
        if self.current_process:
            try:
                self.current_process.terminate()
                self.current_process.wait(timeout=5)
            except subprocess.TimeoutExpired:
                self.current_process.kill()
                self.current_process.wait()
            except:
                pass
            finally:
                self.current_process = None

    def play_video(self, video_path, duration=None):
        """Play video using omxplayer with hardware acceleration"""
        print(f"Playing video: {video_path}")
        
        self.stop_current_playback()
        
        try:
            cmd = [
                'omxplayer',
                '--no-keys',           # Disable keyboard input
                '--no-osd',            # Disable on-screen display
                '--blank',             # Blank screen before playing
                '--aspect-mode', 'fill', # Fill screen
                str(video_path)
            ]
            
            # Add loop option if no duration specified
            if duration is None:
                cmd.insert(-1, '--loop')
            
            self.current_process = subprocess.Popen(
                cmd,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL
            )
            
            # If duration specified, wait and then stop
            if duration:
                time.sleep(duration)
                self.stop_current_playback()
            
        except FileNotFoundError:
            print("omxplayer not found, trying VLC as fallback")
            self.play_video_vlc(video_path, duration)
        except Exception as e:
            print(f"Error playing video: {e}")

    def play_video_vlc(self, video_path, duration=None):
        """Fallback video player using VLC"""
        try:
            cmd = [
                'vlc',
                '--intf', 'dummy',
                '--no-video-title-show',
                '--fullscreen',
                '--no-mouse-events',
                '--no-keyboard-events',
                str(video_path)
            ]
            
            if duration is None:
                cmd.extend(['--loop'])
            
            self.current_process = subprocess.Popen(
                cmd,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL
            )
            
            if duration:
                time.sleep(duration)
                self.stop_current_playback()
                
        except Exception as e:
            print(f"Error playing video with VLC: {e}")

    def play_image(self, image_path, duration=10):
        """Display image using fbi"""
        print(f"Displaying image: {image_path} for {duration}s")
        
        self.stop_current_playback()
        
        try:
            # Resize image if needed
            resized_path = self.resize_image_if_needed(image_path)
            
            cmd = [
                'fbi',
                '-d', '/dev/fb0',
                '-T', '1',
                '-noverbose',
                '-a',  # autozoom
                str(resized_path)
            ]
            
            self.current_process = subprocess.Popen(
                cmd,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL
            )
            
            # Display for specified duration
            time.sleep(duration)
            self.stop_current_playback()
            
        except Exception as e:
            print(f"Error displaying image: {e}")

    def resize_image_if_needed(self, image_path):
        """Resize image if it's too large for optimal display"""
        try:
            # Check if image is already reasonable size
            with Image.open(image_path) as img:
                width, height = img.size
                
                # If image is already reasonable size, return original
                if width <= self.display_width * 1.2 and height <= self.display_height * 1.2:
                    return image_path
                
                # Create resized version
                resized_dir = self.content_dir / 'resized'
                resized_dir.mkdir(exist_ok=True)
                
                resized_path = resized_dir / f"resized_{Path(image_path).name}"
                
                # Don't resize if already exists and is newer
                if (resized_path.exists() and 
                    resized_path.stat().st_mtime > Path(image_path).stat().st_mtime):
                    return resized_path
                
                # Resize maintaining aspect ratio
                img.thumbnail((self.display_width, self.display_height), Image.Resampling.LANCZOS)
                img.save(resized_path, optimize=True, quality=85)
                
                print(f"Resized {image_path} -> {resized_path}")
                return resized_path
                
        except Exception as e:
            print(f"Error resizing image: {e}")
            return image_path  # Return original on error

    def load_playlist(self, playlist_data):
        """Load playlist into queue"""
        print(f"Loading playlist with {len(playlist_data.get('items', []))} items")
        
        # Clear existing queue
        while not self.playlist_queue.empty():
            try:
                self.playlist_queue.get_nowait()
            except:
                break
        
        # Add items to queue
        for item in playlist_data.get('items', []):
            media_info = item.get('media_asset')
            if media_info:
                media_item = {
                    'name': media_info.get('name', 'Unknown'),
                    'url': media_info.get('url'),
                    'mime_type': media_info.get('mime_type'),
                    'duration': item.get('duration', 10),  # Default 10s for images
                    'local_path': None
                }
                self.playlist_queue.put(media_item)
        
        print(f"Playlist loaded with {self.playlist_queue.qsize()} items")

    def download_media_item(self, media_item):
        """Download media item to local storage"""
        if not media_item['url']:
            return None
            
        try:
            import requests
            
            # Create filename from URL
            filename = Path(media_item['url']).name
            if not filename or '.' not in filename:
                # Generate filename from name and mime_type
                ext = self.get_extension_from_mime(media_item['mime_type'])
                safe_name = "".join(c for c in media_item['name'] if c.isalnum() or c in (' ', '-', '_')).strip()
                filename = f"{safe_name}{ext}"
            
            local_path = self.content_dir / filename
            
            # Skip download if file exists and is recent
            if local_path.exists() and time.time() - local_path.stat().st_mtime < 3600:
                media_item['local_path'] = local_path
                return local_path
            
            print(f"Downloading {media_item['name']}...")
            
            response = requests.get(media_item['url'], timeout=30)
            response.raise_for_status()
            
            with open(local_path, 'wb') as f:
                f.write(response.content)
            
            media_item['local_path'] = local_path
            print(f"Downloaded: {local_path}")
            return local_path
            
        except Exception as e:
            print(f"Error downloading {media_item['name']}: {e}")
            return None

    def get_extension_from_mime(self, mime_type):
        """Get file extension from MIME type"""
        mime_map = {
            'image/jpeg': '.jpg',
            'image/png': '.png',
            'image/gif': '.gif',
            'image/webp': '.webp',
            'video/mp4': '.mp4',
            'video/webm': '.webm',
            'video/quicktime': '.mov'
        }
        return mime_map.get(mime_type, '.bin')

    def play_media_item(self, media_item):
        """Play a single media item"""
        if not media_item['local_path'] or not os.path.exists(media_item['local_path']):
            print(f"Media file not found: {media_item['name']}")
            return
        
        file_ext = Path(media_item['local_path']).suffix.lower()
        
        if file_ext in self.video_formats:
            duration = media_item.get('duration')
            # For videos, if duration is specified and > 0, use it, otherwise loop indefinitely
            if duration and duration > 0:
                self.play_video(media_item['local_path'], duration)
            else:
                self.play_video(media_item['local_path'])  # Loop indefinitely
        elif file_ext in self.image_formats:
            duration = media_item.get('duration', 10)  # Default 10s for images
            self.play_image(media_item['local_path'], max(duration, 1))  # Min 1s
        else:
            print(f"Unsupported media format: {file_ext}")

    def start_playlist_playback(self):
        """Start playlist playback loop"""
        if self.is_playing:
            return
        
        def playback_worker():
            print("Starting playlist playback")
            self.is_playing = True
            
            while self.is_playing:
                if self.playlist_queue.empty():
                    print("Playlist empty, waiting...")
                    time.sleep(10)
                    continue
                
                try:
                    # Get next item (this will block if queue is empty)
                    media_item = self.playlist_queue.get(timeout=1)
                    
                    # Download if needed
                    if not media_item['local_path']:
                        local_path = self.download_media_item(media_item)
                        if not local_path:
                            continue
                    
                    # Play the media
                    self.play_media_item(media_item)
                    
                    # Add item back to end of queue for looping
                    self.playlist_queue.put(media_item)
                    
                except queue.Empty:
                    continue
                except Exception as e:
                    print(f"Error in playback loop: {e}")
                    time.sleep(5)
            
            print("Playlist playback stopped")
        
        self.playback_thread = threading.Thread(target=playback_worker, daemon=True)
        self.playback_thread.start()

    def stop_playlist_playback(self):
        """Stop playlist playback"""
        print("Stopping playlist playback")
        self.is_playing = False
        self.stop_current_playback()
        
        if self.playback_thread:
            self.playback_thread.join(timeout=5)

    def show_test_pattern(self):
        """Show a test pattern for debugging"""
        print("Showing test pattern")
        
        # Create simple test pattern
        from PIL import Image, ImageDraw, ImageFont
        
        img = Image.new('RGB', (self.display_width, self.display_height), color='black')
        draw = ImageDraw.Draw(img)
        
        # Draw test pattern
        colors = [(255, 0, 0), (0, 255, 0), (0, 0, 255), (255, 255, 0), (255, 0, 255), (0, 255, 255)]
        bar_width = self.display_width // len(colors)
        
        for i, color in enumerate(colors):
            x1 = i * bar_width
            x2 = (i + 1) * bar_width
            draw.rectangle([x1, 0, x2, self.display_height//2], fill=color)
        
        # Draw text
        try:
            font = ImageFont.truetype('/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf', 60)
        except:
            font = ImageFont.load_default()
        
        text = "MESOPHY TEST PATTERN"
        text_bbox = draw.textbbox((0, 0), text, font=font)
        text_width = text_bbox[2] - text_bbox[0]
        draw.text(((self.display_width - text_width) // 2, self.display_height // 2 + 50), 
                 text, fill=(255, 255, 255), font=font)
        
        # Save and display
        test_path = self.content_dir / 'test_pattern.png'
        img.save(test_path)
        
        self.play_image(test_path, 30)  # Show for 30 seconds

    def cleanup(self):
        """Clean up resources"""
        print("Cleaning up media player...")
        self.stop_playlist_playback()


# Example usage
if __name__ == "__main__":
    player = NativeMediaPlayer()
    
    # Show test pattern
    player.show_test_pattern()
    
    # Example playlist playback
    # playlist = {
    #     'items': [
    #         {
    #             'media_asset': {
    #                 'name': 'Sample Image',
    #                 'url': 'https://example.com/image.jpg',
    #                 'mime_type': 'image/jpeg'
    #             },
    #             'duration': 10
    #         }
    #     ]
    # }
    # 
    # player.load_playlist(playlist)
    # player.start_playlist_playback()
    # 
    # try:
    #     while True:
    #         time.sleep(10)
    # except KeyboardInterrupt:
    #     player.cleanup()

File: pi-client/lib/media-player.js
const { spawn } = require('child_process');
const fs = require('fs-extra');
const path = require('path');
const mime = require('mime-types');

class MediaPlayer {
  constructor(config, displayInfo) {
    this.config = config;
    this.displayInfo = displayInfo;
    this.currentProcess = null;
    this.currentMedia = null;
    this.isPlaying = false;
    this.playbackStartTime = null;
    this.playbackCallback = null;
  }

  async playMedia(mediaFile, duration = null, callback = null) {
    try {
      // Stop any current playback
      await this.stop();
      
      this.currentMedia = mediaFile;
      this.playbackCallback = callback;
      this.playbackStartTime = Date.now();
      
      const mimeType = mime.lookup(mediaFile.local_path) || 'unknown';
      const mediaType = this.getMediaType(mimeType);
      
      console.log(`Playing ${mediaType}: ${mediaFile.name} (${mimeType})`);
      
      switch (mediaType) {
        case 'video':
          await this.playVideo(mediaFile, duration);
          break;
        case 'image':
          await this.playImage(mediaFile, duration);
          break;
        case 'audio':
          await this.playAudio(mediaFile, duration);
          break;
        default:
          throw new Error(`Unsupported media type: ${mimeType}`);
      }
      
      this.isPlaying = true;
      return true;
      
    } catch (error) {
      console.error('Error playing media:', error);
      if (this.playbackCallback) {
        this.playbackCallback(error);
      }
      return false;
    }
  }

  async playVideo(mediaFile, duration = null) {
    const videoPath = mediaFile.local_path;
    
    // Try omxplayer first (best for Pi hardware acceleration)
    if (await this.commandExists('omxplayer')) {
      console.log('Using omxplayer for video playback');
      await this.playWithOmxplayer(videoPath, duration);
    }
    // Fallback to VLC
    else if (await this.commandExists('vlc')) {
      console.log('Using VLC for video playback');
      await this.playWithVLC(videoPath, duration);
    }
    // Final fallback to mpv
    else if (await this.commandExists('mpv')) {
      console.log('Using mpv for video playback');
      await this.playWithMpv(videoPath, duration);
    }
    else {
      throw new Error('No suitable video player found. Install omxplayer, vlc, or mpv.');
    }
  }

  async playWithOmxplayer(videoPath, duration) {
    const args = [
      '--no-osd',          // No on-screen display
      '--no-keys',         // Disable keyboard input
      '--aspect-mode', 'letterbox', // Maintain aspect ratio
      '--vol', '0',        // Start muted (adjust as needed)
      videoPath
    ];

    // Add hardware acceleration if enabled
    if (this.config.media.videoHardwareAcceleration) {
      args.unshift('--hw');
    }

    this.currentProcess = spawn('omxplayer', args, {
      stdio: 'ignore',
      detached: false
    });

    this.setupProcessHandlers(duration || mediaFile.duration);
  }

  async playWithVLC(videoPath, duration) {
    const args = [
      '--intf', 'dummy',              // No interface
      '--no-video-title-show',        // Don't show title
      '--fullscreen',                 // Fullscreen mode
      '--no-mouse-events',            // Disable mouse
      '--no-keyboard-events',         // Disable keyboard
      '--no-osd',                     // No on-screen display
      videoPath,
      'vlc://quit'                    // Auto-quit when done
    ];

    this.currentProcess = spawn('vlc', args, {
      stdio: 'ignore',
      detached: false
    });

    this.setupProcessHandlers(duration || mediaFile.duration);
  }

  async playWithMpv(videoPath, duration) {
    const args = [
      '--no-terminal',        // No terminal output
      '--no-osc',            // No on-screen controls
      '--no-input-default-bindings', // Disable default key bindings
      '--fullscreen',         // Fullscreen mode
      '--loop-file=no',      // Don't loop
      videoPath
    ];

    this.currentProcess = spawn('mpv', args, {
      stdio: 'ignore',
      detached: false
    });

    this.setupProcessHandlers(duration || mediaFile.duration);
  }

  async playImage(mediaFile, duration = null) {
    const imagePath = mediaFile.local_path;
    const displayDuration = duration || this.config.media.defaultImageDuration || 10;

    // Try fbi first (framebuffer image viewer)
    if (await this.commandExists('fbi')) {
      console.log('Using fbi for image display');
      await this.playWithFbi(imagePath, displayDuration);
    }
    // Fallback to fim
    else if (await this.commandExists('fim')) {
      console.log('Using fim for image display');
      await this.playWithFim(imagePath, displayDuration);
    }
    else {
      throw new Error('No suitable image viewer found. Install fbi or fim.');
    }
  }

  async playWithFbi(imagePath, duration) {
    const framebuffer = this.config.system.framebufferDevice || '/dev/fb0';
    
    const args = [
      '-d', framebuffer,      // Framebuffer device
      '-T', '1',              // Use terminal 1
      '-noverbose',           // Quiet output
      '-a',                   // Auto-zoom to fit screen
      '--once',               // Display once
      imagePath
    ];

    this.currentProcess = spawn('fbi', args, {
      stdio: 'ignore',
      detached: false
    });

    // Set up timer to stop after duration
    this.setupProcessHandlers(duration * 1000);
  }

  async playWithFim(imagePath, duration) {
    const framebuffer = this.config.system.framebufferDevice || '/dev/fb0';
    
    const args = [
      '--device', framebuffer, // Framebuffer device
      '--quiet',               // Quiet output
      '--autozoom',           // Auto-zoom to fit
      '--no-etc-fimrc',       // Don't load config
      '--no-commandline',     // Disable command line
      imagePath
    ];

    this.currentProcess = spawn('fim', args, {
      stdio: 'ignore',
      detached: false
    });

    // Set up timer to stop after duration
    this.setupProcessHandlers(duration * 1000);
  }

  async playAudio(mediaFile, duration = null) {
    // For audio files, we can use aplay for WAV or mpg123 for MP3
    const audioPath = mediaFile.local_path;
    const mimeType = mime.lookup(audioPath);
    
    if (mimeType.includes('wav')) {
      if (await this.commandExists('aplay')) {
        console.log('Using aplay for audio playback');
        this.currentProcess = spawn('aplay', [audioPath], {
          stdio: 'ignore',
          detached: false
        });
      }
    } else if (mimeType.includes('mp3')) {
      if (await this.commandExists('mpg123')) {
        console.log('Using mpg123 for audio playback');
        this.currentProcess = spawn('mpg123', ['-q', audioPath], {
          stdio: 'ignore',
          detached: false
        });
      }
    } else {
      // Fallback to VLC for other audio formats
      if (await this.commandExists('vlc')) {
        console.log('Using VLC for audio playback');
        await this.playWithVLC(audioPath, duration);
        return;
      }
    }

    if (this.currentProcess) {
      this.setupProcessHandlers(duration || mediaFile.duration);
    } else {
      throw new Error(`No suitable audio player found for ${mimeType}`);
    }
  }

  setupProcessHandlers(timeoutMs = null) {
    if (!this.currentProcess) return;

    // Handle process exit
    this.currentProcess.on('exit', (code, signal) => {
      console.log(`Media process exited with code ${code}, signal ${signal}`);
      this.handlePlaybackEnd(code === 0 ? null : new Error(`Process exit code: ${code}`));
    });

    this.currentProcess.on('error', (error) => {
      console.error('Media process error:', error);
      this.handlePlaybackEnd(error);
    });

    // Set up timeout if specified
    if (timeoutMs && timeoutMs > 0) {
      setTimeout(() => {
        if (this.currentProcess && this.isPlaying) {
          console.log(`Stopping media after ${timeoutMs}ms timeout`);
          this.stop();
        }
      }, timeoutMs);
    }
  }

  handlePlaybackEnd(error = null) {
    const playbackDuration = this.playbackStartTime ? Date.now() - this.playbackStartTime : 0;
    
    this.isPlaying = false;
    this.currentProcess = null;
    
    console.log(`Playback ended after ${playbackDuration}ms`);
    
    if (this.playbackCallback) {
      this.playbackCallback(error, {
        media: this.currentMedia,
        duration: playbackDuration,
        completedNormally: !error
      });
    }
    
    this.currentMedia = null;
    this.playbackStartTime = null;
    this.playbackCallback = null;
  }

  async stop() {
    if (!this.currentProcess || !this.isPlaying) {
      return;
    }

    console.log('Stopping current media playback');

    try {
      // Try graceful termination first
      this.currentProcess.kill('SIGTERM');
      
      // Wait a bit for graceful exit
      await new Promise(resolve => setTimeout(resolve, 1000));
      
      // Force kill if still running
      if (this.currentProcess && !this.currentProcess.killed) {
        this.currentProcess.kill('SIGKILL');
      }
      
    } catch (error) {
      console.error('Error stopping media process:', error);
    }

    this.handlePlaybackEnd();
  }

  getMediaType(mimeType) {
    if (mimeType.startsWith('video/')) {
      return 'video';
    } else if (mimeType.startsWith('image/')) {
      return 'image';
    } else if (mimeType.startsWith('audio/')) {
      return 'audio';
    }
    return 'unknown';
  }

  async commandExists(command) {
    try {
      const { exec } = require('child_process');
      await new Promise((resolve, reject) => {
        exec(`which ${command}`, (error) => {
          if (error) reject(error);
          else resolve();
        });
      });
      return true;
    } catch (error) {
      return false;
    }
  }

  getStatus() {
    return {
      isPlaying: this.isPlaying,
      currentMedia: this.currentMedia,
      playbackStartTime: this.playbackStartTime,
      playbackDuration: this.playbackStartTime ? Date.now() - this.playbackStartTime : 0
    };
  }
}

module.exports = MediaPlayer;